plugin/grpctrace/interceptor.go

```diff
@@ -0,0 +1,437 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpctrace
+
+// gRPC tracing middleware
+// https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-rpc.md
+import (
+	"context"
+	"io"
+	"net"
+	"regexp"
+
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/otel/api/core"
+	"go.opentelemetry.io/otel/api/correlation"
+	"go.opentelemetry.io/otel/api/key"
+	"go.opentelemetry.io/otel/api/trace"
+)
+
+var (
+	rpcServiceKey  = key.New("rpc.service")
+	netPeerIPKey   = key.New("net.peer.ip")
+	netPeerPortKey = key.New("net.peer.port")
+
+	messageTypeKey             = key.New("message.type")
+	messageIDKey               = key.New("message.id")
+	messageUncompressedSizeKey = key.New("message.uncompressed_size")
+)
+
+const (
+	messageTypeSent     = "SENT"
+	messageTypeReceived = "RECEIVED"
+)
+
+// UnaryClientInterceptor returns a grpc.UnaryClientInterceptor suitable
+// for use in a grpc.Dial call.
+//
+// For example:
+//     tracer := global.Tracer("client-tracer")
+//     s := grpc.NewServer(
+//         grpc.WithUnaryInterceptor(grpctrace.UnaryClientInterceptor(tracer)),
+//         ...,  // (existing DialOptions))
+func UnaryClientInterceptor(tracer trace.Tracer) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
```

Comment:
suggest breaking this up, not because of line length in characters, but instead because of complexity of reading all these params at once.